engine/src/asset/model/mod.rs

```diff
@@ -0,0 +1,289 @@
+        (v["x"], v["y"], v["z"])
+        for v in frames[0]
+    ],
+    [],
+    [
+        (t["points"][0]["vertex_index"], t["points"][1]["vertex_index"], t["points"][2]["vertex_index"])
+        for t in triangles
+    ]
+)
+# UV
+uv = mesh.uv_layers.new(name="UV")
+for loop in mesh.loops:
+    i = loop.index
+    triangle_point = triangles[i // 3]["points"][i % 3]
+    uv.data[i].uv = (triangle_point["u"], triangle_point["v"])
+# Texture
+image = bpy.data.images.new("Texture", texture_width, texture_height)
+image.pixels = texture
+image.update()
+# Material
+material = bpy.data.materials.new(name="Material")
+material.use_nodes = True
+material_bsdf = material.node_tree.nodes["Principled BSDF"]
+material_bsdf.inputs["Roughness"].default_value = 1.0
+material_texture = material.node_tree.nodes.new("ShaderNodeTexImage")
+material_texture.image = image
+material_texture.interpolation = "Closest"
+material.node_tree.links.new(material_texture.outputs["Color"], material_bsdf.inputs["Base Color"])
+mesh.materials.append(material)
+# Shape keys
+shape_keys = []
+for i, f in enumerate(frames):
+    shape_keys.append(object.shape_key_add(name=f"Key {{i}}"))
+    for v_i, v in enumerate(f):
+        shape_keys[i].data[v_i].co = (v["x"], v["y"], v["z"])
+# Actions
+actions = []
+mesh.shape_keys.animation_data_create()
+for i, s in enumerate(sequences):
+    actions.append(bpy.data.actions.new(f"Action {{i}}"))
+    actions[i].use_fake_user = True
+    actions[i].frame_end = len(s["frames"])
+    actions[i].use_frame_range = True
+    mesh.shape_keys.animation_data.action = actions[i]
+    for f_i, frame in enumerate(s["frames"]):
+        for s_i, shape_key in enumerate(shape_keys):
+            shape_key.value = 1.0 if s_i == frame else 0.0
+            shape_key.keyframe_insert(data_path="value", frame=f_i + 1)
+
+# Finalize
+bpy.context.collection.objects.link(object)
+bpy.context.view_layer.objects.active = object
+object.select_set(True)
+            "#,
+                self.texture[0].len(),
+                self.texture.len(),
+                self.texture
+                    .with_palette(&palette)
+                    .into_iter()
+                    // In blender, y axis of textures is reversed
+                    .rev()
+                    .map(|r| format!("    {}", r.into_iter().map(|c| c.to_string()).join(", ")))
+                    .join(",\n"),
+                self.frames
+                    .iter()
+                    .map(|f| format!(
+                        "    [{}]",
+                        f.vertices.iter().map(|v| v.to_string()).join(", ")
+                    ))
+                    .join(",\n"),
+                self.triangles
+                    .iter()
+                    .map(|v| format!("    {v}"))
+                    .join(",\n"),
+                self.sequences
+                    .iter()
+                    .map(|s| format!("    {s}"))
+                    .join(",\n")
+            )
+        }
+    }
```

Comment:
Why???? Just WHY???

Reply:
I added TODOs how to refactor it, which can be done later in a separate PR.